devito/arguments.py

```diff
@@ -175,51 +212,41 @@ def dtype(self):
 
     @cached_property
     def rtargs(self):
-        if self.size is not None:
-            return []
-        else:
-            size = ScalarArgument("%s_size" % self.name, self, max)
-            return [size]
-
-    @property
-    def decl(self):
-        return self.rtargs[0].decl
+        size = ScalarArgument("%s_size" % self.name, self, max)
+        return [size]
 
     # TODO: Can we do without a verify on a dimension?
     def verify(self, value):
         verify = True
+        if value is None:
+            if self.value is not None:
+                return True
 
-        if value is None and self._value is not None:
-            return verify
+            try:
+                value = self.parent.value
+                if value is None:
+                    return False
+            except AttributeError:
+                return False
 
-        if value is not None and value == self._value:
-            return verify
+        try:
+            parent_value = self.parent.value
+            if parent_value is not None:
+                value = self.reducer(value, parent_value)
+            verify = verify and self.parent.verify(value)
+        except AttributeError:
+            pass
 
-        if self.size is not None:
-            # Assuming the only people calling my verify are symbolic data, they need to
-            # be bigger than my size if I have a hard-coded size
-            if not self.is_Buffered:
-                verify = (value >= self.size)
-            else:
-                # If I am a buffered dimension, I just need to make sure the calling
-                # object has enough buffers as my modulo
-                verify = (value >= self.modulo)
-        else:
-            if value is not None and self._value is not None:
-                value = self.reducer(self._value, value)
-            if hasattr(self, 'parent'):
-                verify = verify and self.parent.verify(value)
-                # If I don't know my value, ask my parent
-                if value is None:
-                    value = self.parent.value
-
-            # Derived dimensions could be linked through constraints
-            # At this point, a constraint needs to be added that enforces
-            # dim_e - dim_s < SOME_MAX
-            # Also need a default constraint that dim_e > dim_s (or vice-versa)
-            verify = verify and all([a.verify(v) for a, v in zip(self.rtargs, (value,))])
-            if verify:
-                self._value = value
+        if value == self.value:
+            return True
+
+        # Derived dimensions could be linked through constraints
+        # At this point, a constraint needs to be added that enforces
+        # dim_e - dim_s < SOME_MAX
+        # Also need a default constraint that dim_e > dim_s (or vice-versa)
+        verify = verify and all([a.verify(v) for a, v in zip(self.rtargs, (value,))])
+        if verify:
+            self._value = value
```

Comment:
I still don't like the fact that we are changing state in a method called ``verify``. Is this changing any time soon ?

Reply:
No I haven't worked on that but can we wait until all 3 PRs are in? (This is 1/3)

Reply:
Yes